lib/attack.go

```diff
@@ -263,6 +261,119 @@ func ProxyHeader(h http.Header) func(*Attacker) {
 	}
 }
 
+// DNSCaching returns a functional option that enables DNS caching for
+// the given ttl. When ttl is zero cached entries will never expire.
+// When ttl is non-zero, this will start a refresh go-routine that updates
+// the cache every ttl interval. This go-routine will be stopped when the
+// attack is stopped.
+// When the ttl is negative, no caching will be performed.
+func DNSCaching(ttl time.Duration) func(*Attacker) {
+	return func(a *Attacker) {
+		if ttl < 0 {
+			return
+		}
+
+		if tr, ok := a.client.Transport.(*http.Transport); ok {
+			dial := tr.DialContext
+			if dial == nil {
+				dial = a.dialer.DialContext
+			}
+
+			resolver := &dnscache.Resolver{}
+
+			if ttl != 0 {
+				go func() {
+					refresh := time.NewTicker(ttl)
+					defer refresh.Stop()
+					for {
+						select {
+						case <-refresh.C:
+							resolver.Refresh(true)
+						case <-a.stopch:
+							return
+						}
+					}
+				}()
+			}
+
+			rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+			tr.DialContext = func(ctx context.Context, network, addr string) (conn net.Conn, err error) {
+				host, port, err := net.SplitHostPort(addr)
+				if err != nil {
+					return nil, err
+				}
+
+				ips, err := resolver.LookupHost(ctx, host)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(ips) == 0 {
+					return nil, &net.DNSError{Err: "no such host", Name: addr}
+				}
+
+				// Pick a random IP from each IP family and dial each concurrently.
+				// The first that succeeds wins, the other gets canceled.
+
+				rng.Shuffle(len(ips), func(i, j int) { ips[i], ips[j] = ips[j], ips[i] })
+
+				// In place filtering of ips to only include the first IPv4 and IPv6.
+				j := 0
+				for i := 0; i < len(ips) && j < 2; i++ {
+					ip := net.ParseIP(ips[i])
+					switch {
+					case len(ip.To4()) == net.IPv4len && j == 0:
+						fallthrough
+					case len(ip) == net.IPv6len && j == 1:
+						ips[j] = ips[i]
+						j++
+					}
+				}
+				ips = ips[:j]
+
+				type result struct {
+					conn net.Conn
+					err  error
+				}
+
+				ch := make(chan result, len(ips))
+				ctx, cancel := context.WithCancel(ctx)
+				defer cancel()
+
+				for _, ip := range ips {
+					go func(ip string) {
+						conn, err := dial(ctx, network, net.JoinHostPort(ip, port))
+						ch <- result{conn, err}
+					}(ip)
+				}
+
+				for i := 0; i < cap(ch); i++ {
+					select {
+					case <-ctx.Done():
+						return nil, ctx.Err()
+					case r := <-ch:
+						if err = r.err; err != nil {
+							continue
+						}
+						return r.conn, nil
+					}
+				}
+
+				return nil, err
+			}
+		}
+	}
+}
+
+type attack struct {
+	name  string
+	began time.Time
+
+	seqmu sync.Mutex
+	seq   uint64
+}
```

Comment:
If seqmu protects only seq uint64, why not use e.g. [sync/atomic.Uint64](https://pkg.go.dev/sync/atomic#Uint64)?

Reply:
Because it needs to be used to not only protect an increment to seq but also make sure the associated timestamps are in the same order as the sequence number: https://github.com/tsenart/vegeta/blob/9fba76955e42c37f86c7a1c3ebde3b529a145afb/lib/attack.go#L485-L489

Reply:
Ah, so given Result values r1 and r2, r1.Seq > r2.Seq implies r1.Timestamp.After(r2.Timestamp)? Very subtle!

Reply:
That's right. Subtle indeed!